homework_2_function/fastq_filtrator.py

```diff
@@ -0,0 +1,107 @@
+# GC-content calculator.
+# reads - string, DNA sequence.
+def count_gc_content(reads):
+    reads = list(reads)
+    return (reads.count('G') + reads.count('C')) / len(reads) * 100
+
+
+# Creates a list with an interval for filtering.
+# interval - tuple, int or float.
+def create_interval(interval):
+    if type(interval) == tuple:
+        return list(interval)
+    elif (type(interval) == int) | (type(interval) == float):
+        interval_list = [0, interval]
+        return interval_list
+    else:
+        print('Not supported types')
+
+# Mean quality calculator.
+# read_quality - string, score phred33.
+def medium_quality(read_quality):
+    read_quality = list(read_quality)
+    sum_quality = 0
+    for i in read_quality:
+        sum_quality += ord(i) - 33
+    return sum_quality / len(read_quality)
+
+
+# Reads specific line numbers.
+# path_to_file - string, from start to end
+def read_fastq(path_to_file, start=0, end=100):
+    with open(path_to_file) as file:
+        lines = file.readlines()
+    return lines[start:end]
+
+
+# Counts how many lines are in a file.
+# path_to_file - string.
+def length_file(path_to_file):
+    with open(path_to_file, 'r') as file:
+        count_lines = sum(1 for line in file)
+    return count_lines
+
+
+# The main function for filtering by quality, length and GC-content.
+# input_fastq, output_file_prefix - strings
+# gc_bounds, length_bounds - tuple, int or float
+# quality_threshold - int or float
+# save_filtered - bool
+def main(input_fastq,
+         output_file_prefix,
+         gc_bounds=(0, 100),
+         length_bounds=(0, 2**32),
+         quality_threshold=0,
+         save_filtered=False):
+
+    # Empty list to add reads.
+    good_list = []
+    bad_list = []
+
+    # List with an interval for filtering.
+    gc_bounds = create_interval(gc_bounds)
+    length_bounds = create_interval(length_bounds)
+
+    start_line = 0
+
+    # Filtering loop
+    while start_line < length_file(input_fastq):
+
+        if gc_bounds[1] < gc_bounds[0]:
+            print('Attention! Wrong interval for GC-content.')
+            break
+        if length_bounds[1] < length_bounds[0]:
+            print('Attention! Wrong interval for length.')
+            break
+
+        # List of 4 lines from .fastq file
+        read_seq = read_fastq(input_fastq, start_line, start_line+4)
+
+        gc_test = True
+        quality_test = True
+        len_test = True
+
+        # Conditions for filtering
+        if (count_gc_content(read_seq[1]) < gc_bounds[0]) | (count_gc_content(read_seq[1]) > gc_bounds[1]):
+
+            gc_test = False

```

Comment:
1. Зачем вызывать функцию два раза, если можно вызвать один
2. Выражение можно упростить
3. Скомбинируем 1 и 2 
```suggestion
        gc_test =  gc_bounds[0] <= count_gc_content(read_seq[1]) <= gc_bounds[1]
```